cli/context/store/io_utils.go

```diff
@@ -0,0 +1,56 @@
+package store
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+)
+
+// limitedReader a wrapper on io.Reader for error handling of ReadAll
+type limitedReader struct {
+	r     io.Reader
+	limit int64
+	n     int64
+	err   error
+}
+
+// newLimitedReader will result in a limited reader with defined byte limit.
+// This basically extends io.LimitReader with proper errors as io.LimitReader only errors with EOF.
```

Comment:
io.LimitReader does not only error with EOF. https://golang.org/src/io/io.go#L448

I would much rather use io.LimitReader. Unless I'm misunderstanding something.

Reply:
OK I understand now, what you want is https://golang.org/pkg/net/http/#MaxBytesReader but with `io` instead of `http`.

Reply:
I think I wasn't clear in the comment there. I think the main concern here from @thaJeztah & me was the fact that, if reader hits the limit it won't tell us it's actually over the limit, but instead will just return an `EOF`.

Reply:
Ah, yes, the comment in there `returns a non-EOF error for a Read beyond the limit` is basically what we wanted.

Reply:
Ok so you can simplify this further, because all you need is just a 2 lines change around https://golang.org/src/io/io.go?s=14885:14942#L443
```Go
type LimitedReader struct {
	io.LimitedReader
}

func (l *LimitedReader) Read(p []byte) (n int, err error) {
	if l.N < 0 {
		return 0, errors.New("read limit exceeded")
	}
	if l.N == 0 {
		return 0, io.EOF
	}
	if int64(len(p)) > l.N {
		p = p[0:l.N]
	}
	n, err = l.R.Read(p)
	l.N -= int64(n)
	return
}
```

This should possibly in its own ioutils package, although I see we have on in docker/docker/pkg/ioutils.
Maybe we keep it here for now (with the risk of forgetting about it forever).

Reply:
You don't need limitedReadAll. Just call ioutil.ReadAll inline.

Reply:
I've implemented the `Read` as discussed, and called `LimitedReader` inline from all the places used for context import. I'll resolve this conversation just for the sake of the iteration causing the code being outdated. Please feel free to open another thread if need be.